Add Drop-down for Analytics vendors in AMP setting page

rutviksavsani opened the pull request:
## Summary

<!-- Please reference the issue(s) this PR fixes, if one exists. For significant changes, opening an issue first for discussion is usually preferable. -->
Fixes #6367 
This PR replaces the text input with a drop-down for "Analytics vendors" on the AMP setting page. PR also adds new CLI command to fetches all the vendors from the [analytics-vendors-list.md](https://github.com/ampproject/amphtml/blob/main/extensions/amp-analytics/analytics-vendors-list.md) and generate a JSON file.

## Screenshots and Videos: 

![Vendors list](https://user-images.githubusercontent.com/53530700/156742807-e1c74b7f-e0b1-4b6d-bbc0-8f1ad8ae781b.png)

https://user-images.githubusercontent.com/53530700/156742827-fbb6d602-ac73-4b18-b82b-b7a63220d33c.mp4

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

rutviksavsani committed:
Add CLI command to generate analytics vendor file in JSON format.

rutviksavsani committed:
Add JSON file of analytic vendors.

rutviksavsani committed:
Add dropdown menu for analytic vendors in AMP setting page instead of textbox.

rutviksavsani committed:
Add unit test cases.

rutviksavsani committed:
Add E2E test cases.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ampproject/amp-wp?pullRequest=6894) <br/>All committers have signed the CLA.

delawski commented on bin/update-analytics-vendors.js:
```diff
@@ -0,0 +1,143 @@
+	async fetchData() {
+		// Remote URL for analytics vendors.
+		const url =
+			'https://api.github.com/repos/ampproject/amphtml/contents/extensions/amp-analytics/analytics-vendors-list.md';
```
Is there a reason we can't use the `md` file directly, i.e.:

```
https://raw.githubusercontent.com/ampproject/amphtml/main/extensions/amp-analytics/analytics-vendors-list.md
```

I guess we wouldn't have to base64 decode file contents from a JSON file in such a case.

rutviksavsani replied:
This suggestion seems more helpful and reduces the extra steps of decoding than the current implementation.

rutviksavsani replied:
resolved in 35b1373041d70dbfd2e3fad7519eb9564ad6aed7

delawski commented on data/analytics-vendors-list.json:
```diff
@@ -0,0 +1,94 @@
+    "gfksensic": "GfK Sensic",
+    "googleadwords": "Google Ads",
+    "googleanalytics": "Google Analytics",
+    "gtag": "gtag",
+    "N/A": "Google Tag Manager",
```
Is this correct?

rutviksavsani replied:
Yes, It is referenced from [analytics-vendors-list.md](https://github.com/ampproject/amphtml/blob/main/extensions/amp-analytics/analytics-vendors-list.md#google-tag-manager).
We can also change it to an empty string, Is there anything else that you suggest?

delawski replied:
I think that an empty string would collide with the initial empty option i.e. "Select Type…" (they both would have a value of `''`).

rutviksavsani replied:
Agree, but I guess passing a `''` string will be more feasible, The [documentation](https://github.com/ampproject/amphtml/blob/main/extensions/amp-analytics/analytics-vendors-list.md#google-tag-manager) suggests keeping it empty. Also, will there be an issue if two options have the same value?

rutviksavsani replied:
@delawski  What we can do is let `Google Tag Manager` be as a default instead of "Select Type…" and hence we would have only one Option with `''` value.

delawski commented on data/analytics-vendors-list.json:
```diff
@@ -0,0 +1,94 @@
+    "umenganalytics": "Umeng+ Analytics",
+    "upscore": "Upscore",
+    "vponanalytics": "Vpon Analytics",
+    "webengage.": "Webengage",
```
I think the period should not be part of the vendor slug.

rutviksavsani replied:
Yes, It's a miss from my end will make the changes

rutviksavsani replied:
resolved in 35b1373041d70dbfd2e3fad7519eb9564ad6aed7

delawski commented on bin/update-analytics-vendors.js:
```diff
@@ -0,0 +1,143 @@
+	 *
+	 * @return {Promise<void>}
+	 */
+	async extractVendors() {
```
I stumbled upon a [markdown parser](https://github.com/markedjs/marked) that may be handy if the manual parsing turns out to be too fragile.

rutviksavsani replied:
There is not much fragile manual parsing as of now, So I guess we won't require this parser.

delawski commented on assets/src/settings-page/analytics.js:
```diff
@@ -116,6 +117,13 @@ function AnalyticsEntry( { entryIndex, onChange, onDelete, type = '', config = '
+	[ ...Object.keys( AnalyticsVendors ) ].forEach( ( vendor ) => {
+		options.push( { value: vendor, label: AnalyticsVendors[ vendor ] } );
+	} );
+
```
Since this does not depend on the component's state and only on external data (`AnalyticsVendors`), it should be moved out of the component body so that the `option` are not recalculated on every render.

Also, the empty option label should be translatable, i.e. `__( 'Select Type…', 'amp' )`.

Lastly, `Object.entries` could be used for generating the `options` array like:

```js
const options = [
	{ value: '', label: __( 'Select Type…', 'amp' ) },
	...Object.entries( AnalyticsVendors ).map( ( [ value, label ] ) => ( { value, label } ) )
];
```

rutviksavsani replied:
Resolved in 27e4f23bd95a41a323337a117a4c93d0f74a2d67

delawski reviewed (changes requested):
Thank you for working on this feature. I've left a few comments in the code.

delawski commented:
Note that the CSS lint issues seem to be unrelated to the changes in this PR.

rutviksavsani committed:
Update URL for Analytics Aendors and fix extra characters in JSON

rutviksavsani committed:
Rename value constant and Optimize options constant for dropdown menu.

rutviksavsani committed:
Change key for Googele Tag Manager in analytics vendors to empty string

rutviksavsani committed:
Remove Select Type... option from analytics vendors dropdown

delawski commented on assets/src/settings-page/analytics.js:
```diff
@@ -111,11 +112,16 @@ function AnalyticsEntry( { entryIndex, onChange, onDelete, type = '', config = '
 		? defaultValue
 		: config;
 
+	/* Populating options for analytics vendors. */
+	const options = [
+		...Object.entries( AnalyticsVendors ).map( ( [ value, label ] ) => ( { value, label } ) ),
+	];
+
```
Two things about this part:

1. There's no need to use an array spread now. `Object.entries()` already returns an array and we're just `map`ping over it so the resulting value is an array all of the time. (Earlier, since we've been prepending an empty option to the `options` array, we had to use an array spread.) 

```suggestion
	/* Populating options for analytics vendors. */
	const options = Object.entries( AnalyticsVendors ).map( ( [ value, label ] ) => ( { value, label } ) );

```

2. This code should be moved outside of the functional component, i.e. outside of the `AnalyticsEntry` so that we're not recreating the `options` array on each component render. The `AnalyticsVendors` object is "static" (it's imported from a JSON file) so it doesn't depend on a component state and so we can do this mapping only once.

rutviksavsani replied:
Resolved the issue.

delawski reviewed (changes requested):
Thanks for addressing the feedback @rutviksavsani. There's one thing left to do, please check it out.

rutviksavsani committed:
Remove Analytics vendors options from functional scope

rutviksavsani commented:
Thanks, @delawski for patiently looking into the PR.

delawski reviewed (approved):
@rutviksavsani Thanks for following up.

@westonruter This looks good to me, passing over to you.

westonruter commented on assets/src/settings-page/analytics.js:
```diff
@@ -125,18 +129,15 @@ function AnalyticsEntry( { entryIndex, onChange, onDelete, type = '', config = '
 						} }
-						type="text"
-						pattern="^[a-zA-Z0-9_.-]+$"
-						title={ __( 'Must be a valid vendor or left blank for in-house analytics.', 'amp' ) }
```
It seems the new dropdown doesn't allow for the an empty vendor to be supplied? The ability to specific an empty vendor is needed still for in-house analytics.

rutviksavsani replied:
There is confusion regarding this, also discussed [here](https://github.com/ampproject/amp-wp/pull/6894#discussion_r802854794) 
both `Google Tag Manager` and in-house analytics expect a `''` empty string which when adding to the dropdown conflicts in value and the first option with empty value gets selected.

What do you suggest for resolving this?

rutviksavsani replied:
@westonruter @delawski I need your suggestions for the above comment.

delawski replied:
@rutviksavsani That's a good question. I think we have two possibilities here:

- Use an empty option labelled like "None" or "---" as the first item on the list and provide a help note that for Google Tag Manager an empty vendor type should be selected.
- Instead of "None" or "---", use something like: "In house analytics / Google Tag Manager" for the empty option label.

@westonruter What are your thoughts?

rutviksavsani replied:
@delawski 
The second suggestion seems a more approachable way to resolve the confusion and also informative for the user at the same time without using any extra elements.
@westonruter what do you say?

westonruter commented on assets/src/settings-page/analytics.js:
```diff
@@ -66,6 +67,9 @@ const vendorConfigs = {
 };
 
+/* Populating dropdown options by analytics vendors. */
+const AnalyticsVendorsOptions = Object.entries( AnalyticsVendors ).map( ( [ value, label ] ) => ( { value, label } ) );
```
Why not store the entries in this way initially? Then there would be no need to do this conversion every time the page loads.

westonruter replied:
Also, the array should be sorted alphabetically by the `label` in the data that is stored.

rutviksavsani replied:
Now, The entries are being stored in the required manner initially in the array.

westonruter commented on src/Admin/OptionsMenu.php:
```diff
@@ -185,6 +185,24 @@ public function screen_handle() {
+			return []; // @codeCoverageIgnore
+		}
+
+		$vendors_list = file_get_contents( $vendors_list_file );
+		$vendors_list = json_decode( $vendors_list, true );
```
Instead of storing the file as JSON, it could instead be stored as a PHP array in the same way that `update-extension-files.js` is doing. This would be faster to load.

rutviksavsani replied:
Agree, I did follow the Issue [description](https://github.com/ampproject/amp-wp/issues/6367#issuecomment-859030499)
for the reference and implemented the JSON file.

Will surely make the changes as suggested now.

westonruter replied:
Case in point: https://twitter.com/schlessera/status/1492250702000832514?t=S4lnlmBMoFPB-ou24kjEug&s=19

rutviksavsani replied:
Resolved
Converted it to a PHP Array

westonruter commented on bin/update-analytics-vendors.js:
```diff
@@ -0,0 +1,139 @@
+/**
+ * File path of the analytics vendors list.
+ */
+const ANALYTICS_VENDORS_FILE = 'data/analytics-vendors-list.json';
```
Let's put this rather in `includes` alongside the `includes/ecosystem-data` directory.

rutviksavsani replied:
Resolved, added it to `includes` as `includes/data/analytics-vendors-list.php`

westonruter commented on assets/src/settings-page/analytics.js:
```diff
@@ -125,18 +129,15 @@ function AnalyticsEntry( { entryIndex, onChange, onDelete, type = '', config = '
-						title={ __( 'Must be a valid vendor or left blank for in-house analytics.', 'amp' ) }
-						placeholder={ __( 'Vendor or blank', 'amp' ) }
 						value={ type }
+						options={ AnalyticsVendorsOptions }
```
The `AnalyticsVendorsOptions` variable name makes it look like a class/component. Should it rather be `analyticsVendorsOptions`?

rutviksavsani replied:
Yes, it should be as you suggested, will surely make the changes.

rutviksavsani replied:
resolved

westonruter reviewed (changes requested):
Thanks for the PR!

rutviksavsani committed:
Convert analytics vendors to PHP array from JSON,Move vendors to includes dir and Fix AnalyticsVendorsOptions const

rutviksavsani committed:
sort and store vendors data initially, Also apend label In house analytics to empty option

rutviksavsani commented:
@westonruter @delawski I have Implemented the requested changes and re-requesting a review for the same.

rutviksavsani committed:
Fix ES list errors in update-analytics-vendors.js

delawski commented on assets/src/settings-page/analytics.js:
```diff
@@ -3,14 +3,15 @@
  */
 import PropTypes from 'prop-types';
 import { v4 as uuidv4 } from 'uuid';
+import { ANALYTICS_VENDORS_LIST as AnalyticsVendors } from 'amp-settings';
```
I think we should follow the same convention as in other JS files, i.e. use the uppercase external constant names. Also, as I believe @westonruter noted earlier, using title case for a variable name (`AnalyticsVendors`) may suggest it is a React component while it's not.

rutviksavsani replied:
@delawski That means you are suggesting it should be `ANALYTICSVENDORS` right?

rutviksavsani replied:
@delawski It is resolved. 
Will take care of this in the future.

delawski replied:
Looks good, thanks!

rutviksavsani committed:
Modify convention for AnalyticsVendors const

rutviksavsani commented:
@westonruter Any inputs here or is it good to go?

westonruter commented:
In `develop` I tried adding analytics entries like the following:

> <img width="939" alt="image" src="https://user-images.githubusercontent.com/134745/157354235-8060eeb8-147b-401c-ac5c-f3f3bbaea6fe.png">

When I switched to this branch, they all unexpectedly were set to match Acquia Lift:

> <img width="936" alt="image" src="https://user-images.githubusercontent.com/134745/157354318-1e8efff1-27d5-4a7e-8cd7-cb1702e6bc98.png">

It is critical that the data format remain compatible so that the values remain consistent when upgrading/downgrading.

westonruter commented on includes/data/analytics-vendors-list.php:
```diff
@@ -0,0 +1,473 @@
+<?php // phpcs:disable Squiz.Commenting.FileComment.Missing
```
Can this file be placed at `includes/ecosystem-data/analytics-vendors.php`?

rutviksavsani replied:
Moved it to the `ecosystem-data` directory.

westonruter commented on bin/update-analytics-vendors.js:
```diff
@@ -0,0 +1,185 @@
+					// Loop through multiple vendor slugs with same titles and append extra information to title.
+					vendorSlugs.forEach( ( slug ) => {
+						if ( vendorSlugs.indexOf( slug ) === 0 ) {
+							// If vendor is Google Tag Manager, then add it along with In house analytics for '' value.
+							if ( slug === 'N/A' && vendorTitle === 'Google Tag Manager' ) {
+								slug = '';
+								vendorTitle = 'In house analytics / Google Tag Manager';
+							}
```
I think this may be too difficult to find in the list. For the other thing, “In house analytics” should really be translated for the user, so I don't think it should be here. 

What's more is that this UI has never really been suitable for GTM because it requires adding a `config` attribute:

```html
<amp-analytics config="https://www.googletagmanager.com/amp.json?id=GTM-CONTAINER_ID" data-credentials="include"></amp-analytics>
```

Whereas the UI here is all about managing the `type` attribute.

Therefore, I might suggest that instead of having this GTM option, that there be another value added to the beginning/end of the list like “(Other)” and when that is selected, a paragraph with information about in-house analytics and GTM can be added. For GTM specifically, we'd recommend rather that the user use Site Kit instead (which [outputs GTM](https://github.com/google/site-kit-wp/blob/ef0cc0ad2a03c6b6b0ad3f2e72b46902274194fa/includes/Modules/Tag_Manager/AMP_Tag.php#L68-L73)), in the same way we provide this guidance if the user supplies `googleanalytics` as the type.

rutviksavsani replied:
@westonruter I have introduced the option as `Other` along with the displayed description and also resolved the saving option issue too.  

![image](https://user-images.githubusercontent.com/53530700/157490643-0b051422-115a-4aa6-b288-1987c71efe4d.png)

rutviksavsani committed:
Change analytics vendors data directory to ecosystem data and introduce new option for GTM and in-house analytics combined.

westonruter commented on assets/src/settings-page/analytics.js:
```diff
@@ -125,18 +134,15 @@ function AnalyticsEntry( { entryIndex, onChange, onDelete, type = '', config = '
-					<TextControl
-						className="option-input"
+				<div className="amp-analytics-entry__select-control">
+					<SelectControl
```
Just had an idea: what if instead of a `SelectControl` this instead implemented a [`datalist`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/datalist)? It appears to be [supported](https://caniuse.com/datalist) by all current browsers. This is a relatively new HTML input type which I have rarely used, but it could be a great fit here. It would be more of a progressive enhancement with the current `input` field, gracefully falling back to just a plain text input for browsers that don't support it (by chance). Also, critically, it would still allow custom entries to be provided, such as if a new analytics vendor is added to AMP before a plugin update is done.

Often a site owner may be told to add `amp-analytics` with a given `type` value and not be told necessarily what the name of the analytics provider is. It would allow a user to type part of the analytics provider's name or part of the `type` and get auto-completion. 

Example:

```html
<label for="vendor">Vendor:</label>
<input list="vendors" id="vendor" name="vendor">
<datalist id="vendors">
  <option value="acquialift">Acquia Lift</option>
  <option value="afsanalytics">AFS Analytics</option>
  <option value="googleanalytics">Google Analytics</option>
</datalist>
```

<img width="283" alt="image" src="https://user-images.githubusercontent.com/134745/157721437-23868750-05eb-439f-992b-5880c88ba2b2.png">

<img width="275" alt="image" src="https://user-images.githubusercontent.com/134745/157721474-9d74eb3f-03bd-44d3-bbec-ee49f78d170e.png">

<img width="275" alt="image" src="https://user-images.githubusercontent.com/134745/157721512-e53dd8bc-599a-4ded-8b05-dafae2514f00.png">

<img width="229" alt="image" src="https://user-images.githubusercontent.com/134745/157721577-d0bcf1a8-d21b-4e7f-9c88-3bd243ac8c32.png">

rutviksavsani replied:
@westonruter I did carry out some findings on the scope of using `datalist` HTML component.

This sounds good to me :+1: 

The only query I need to clarify here is How will we handle datalist for `in-house analytics` and `Google Tag Manager`?

Also, Should we store the analytics vendors as an array of `value` and `Label` and process them later 
 or initially save as an option element in PHP array. 

`<option value="adobeanalytics_nativeConfig"> Adobe Analytics (native config) </option> ` ?

westonruter replied:
> The only query I need to clarify here is How will we handle datalist for `in-house analytics` and `Google Tag Manager`?

That would be as it was before, with the author leaving it empty.

In reality, this UI should not be used for GTM anyway because it is supposed to include `data-credentials="include"` which we don't add. So if someone tries to enter `googleanalytics` we show the notice about using Site Kit. The general information about GTM and in-house analytics could be added to the `details` that appears at the top of the section. 

We _could_ add a `googletagmanager` vendor to the list but then if the user selects it we could disable the field and show a notice in that case, but that's perhaps over the top.

> Also, Should we store the analytics vendors as an array of `value` and `Label` and process them later
> or initially save as an option element in PHP array.

Yes! You got it.

rutviksavsani replied:
@westonruter I dug a bit deep into the Datalist thing and it is not yet customizable by the browsers. 
The issues we may face are 
    1) The `datalist` tag is not customizable by the CSS and is handled by the browser.
    2) We can fix the position and length of `datalist` directly from the HTML or CSS. 
    3) There will be UI changes to datalist as per the browser being used.
  
 Personally, I do find a bit of inconsistency in the visual effect of datalist on the input elements.
  
  ### Attached below is a video to show the behavior of datalist on different browsers.

https://user-images.githubusercontent.com/53530700/158396888-726f573c-7e7a-49c5-8833-02c1d0224622.mp4

Please let me know your view on this and if I should really proceed with the implementation.

rutviksavsani replied:
@westonruter I dug a bit deep into the Datalist element while implementing it and found that it is not yet easily customizable by the developers using the CSS. We will need to use JS to customize it. 
The points to consider are
    1) The `datalist` tag is not customizable by the CSS and is handled by the browser.
    2) We can't fix the position and number of rows of `datalist` directly from the HTML or CSS. 
    3) There will be UI changes to the datalist depending on the browser being used.
  
 Personally, I do find a bit of inconsistency in the visual effect of datalist on the input element.
 
 **Please let me know your view on this and if I should proceed with the implementation.**
 
### Attached below is a video to show the behavior of datalist on different browsers.
https://user-images.githubusercontent.com/53530700/158396888-726f573c-7e7a-49c5-8833-02c1d0224622.mp4

westonruter replied:
@rutviksavsani Thanks for the helpful video. 

I'm not troubled by the lack of styling or UI inconsistencies between browsers. Actually, it was just announced that all the browser vendors have gotten together on the [Interop 2022](https://web.dev/interop-2022/) initiative. They'll be working on fifteen areas of interoperability, and one of these is **forms**. So we should be seeing the consistency between browsers increasing this year. So I think you should go ahead with implementation of datalust.

rutviksavsani replied:
Resolved the changes in https://github.com/ampproject/amp-wp/pull/6894/commits/e1e6f198e9d21886997f41d5d0a2d3aa1c88e14d

rutviksavsani committed:
Fix sample value for vendorConfigs empty key.

rutviksavsani committed:
Replace selectcontrol component in analytics settings with input datalist.

rutviksavsani commented:
@westonruter The requested changes have been implemented.
The changes include :
1) Replaced `SelectControl` component with HTML `Input Datalist` component.
2) Removed the `Other` and `Google Tag Manager` options from the list of analytics vendors and it will not be shown in datalist as well.  If a user enters `googletagmanager` the notice of googleanalytics will be shown. 
3) Modified the Main section instruction as well as the notice keeping in mind `googletagmanger` vendor.

### Modified notice when user enters `googletagmanager`, `googleanalytics` or `gtag`. 
![image](https://user-images.githubusercontent.com/53530700/159247147-73f01cfd-c3b4-4913-9772-9c7eccd2a689.png)

### Modified details for the top of the analytics section.
![image](https://user-images.githubusercontent.com/53530700/159247222-2eebaccb-e878-4957-9b05-92fe38dcf930.png)

### Working of datalist

https://user-images.githubusercontent.com/53530700/159247565-b0771421-35c4-4c75-b2e3-4a2dc5075467.mp4

rutviksavsani committed:
Fix CSS and JS lint errors.

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into enhancement/6367-dropdown-for-analytics-type

* 'develop' of github.com:ampproject/amp-wp: (301 commits)
  Bump minimist from 1.2.5 to 1.2.6
  Bump ansi-regex from 3.0.0 to 3.0.1 in /.github/actions/draft-release
  Bump minimist from 1.2.5 to 1.2.6 in /.github/actions/draft-release
  Update branch for google-github-actions/setup-gcloud
  Prevent BlockUniqidTransformer from running on WP>=5.9.3
  Bump node-forge from 1.2.1 to 1.3.0
  Centralize resetting wp_the_query to wp_query in set_up
  Bump @babel/core from 7.17.7 to 7.17.8
  Harmonize workarounds until cause is found
  Ensure title-tag theme support is present
  Remove code that was apparently made unnecessary by 774050f
  Increase memory limit for phpstan in lint-staged; remove unused PHP imports
  Ensure _wp_theme_features global is reset after each test
  Add changed-php-count check to PCOV setup step
  Restore testing PHP 7.3 and 7.4 w/ external-http
  Added some test cases for AMP_Post_Meta_Box
  Fix syntax error
  Move setup PCOV before copying WP plugin directory
  Add step to setup code coverage module PCOV
  Ensure CSS transients disabling is not reset with each version update
  ...

westonruter commented on src/Admin/OptionsMenu.php:
```diff
@@ -186,6 +186,24 @@ public function screen_handle() {
+			return []; // @codeCoverageIgnore
+		}
+
+		// Fetches analytics vendors list as a array.
+		$vendors_list = require $vendors_list_file;
+
+		$error = json_last_error();
+
+		return ! empty( $error ) ? [] : $vendors_list;
```
Just realized that the JSON code is obsolete since a PHP file is now being written.

```suggestion
		return require $vendors_list_file;
```

westonruter replied:
Maybe a bit easier on the eyes:

```suggestion
		$vendors_list = require $vendors_list_file;

		return $vendors_list;
```

westonruter committed:
Use for/of loop instead of forEach

westonruter committed:
Remove obsolete JSON logic

westonruter committed:
Flesh out get_analytics_vendors test

westonruter commented on assets/src/settings-page/analytics.js:
```diff
@@ -125,19 +140,25 @@ function AnalyticsEntry( { entryIndex, onChange, onDelete, type = '', config = '
+							onChange( { type: newType.target.value } );
+						} }
 					/>
+					<datalist id="vendors" className="input-datalist" >
```
The `vendors` ID is being duplicated here.

westonruter committed:
Fix duplicated vendors ID

westonruter reviewed (approved):
This is a great progressive enhancement! A much-improved way to discover vendor IDs and yet still provide flexibility.

westonruter commented:
@rutviksavsani Thanks for your persistence and patience.
